File: leetcode/098isbst/iterative_answer.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        if not root:
            return True
        
        stack, prev = [], None
        
        while root or len(stack) != 0:
            # Go all the way to the left
            while root:
                stack.append(root)
                root = root.left
            
            # Get leftmost node
            root = stack.pop()            
            
            # If there was a previous node (will always be the previous one in inorder trav)
            if prev and root.val <= prev.val:
                return False
            
            # Prev is now this node, go right one
            # If right is none, we get the parent from line 22
            prev = root
            root = root.right
        
        return True
            

File: leetcode/0145/iterative_answer.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<pair<TreeNode*,int>> s;
        s.push({root, 0});
        while (!s.empty()) {
            TreeNode* cur = s.top().first;
            int state = s.top().second;
            s.pop();
            if (cur == NULL) continue;
            if (state < 2) s.push({cur, state+1});
            if (state == 0) s.push({cur->left, 0});
            else if (state == 1) s.push({cur->right, 0});
            else if (state == 2) res.push_back(cur->val);
        }
        return res;
    }
};
